corporate/lib/stripe.py

```diff
@@ -955,6 +1111,15 @@ def do_change_plan_type(self, *, tier: Optional[int], is_sponsored: bool = False
             raise AssertionError("Unexpected tier")
         do_change_realm_plan_type(self.realm, plan_type, acting_user=self.user)
 
+    @override
+    def process_downgrade(self, plan: CustomerPlan) -> None:
+        from zerver.actions.realm_settings import do_change_realm_plan_type
+
+        assert plan.customer.realm is not None
+        do_change_realm_plan_type(plan.customer.realm, Realm.PLAN_TYPE_LIMITED, acting_user=None)
+        plan.status = CustomerPlan.ENDED
+        plan.save(update_fields=["status"])
```

Comment:
This can possibly be moved to the base `BillingSession` by just calling `self.do_change_plan_type` and adding a  `BillingSession.DOWNGRADE_PLAN_TYPE` value that all the subclasses will set that equals `Realm.PLAN_TYPE_LIMITED` for `RealmBillingSession`. But I guess it's only a few lines to duplicate if we don't do that.